### PR TITLE
WIP - DON'T MERGE - Reduce DDataShardCoordinator memory usage during unreachability (#25444)

### PR DESCRIPTION
**Pull request by fsat**

### CODE REVIEW

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 1]
```diff
@@ -929,6 +929,9 @@ class DDataShardCoordinator(typeName: String, settings: ClusterShardingSettings,
     if (rememberEntities) Set(CoordinatorStateKey, AllShardsKey) else Set(CoordinatorStateKey)
 
   var shards = Set.empty[String]
+
+  var getShardHomeRequests: Set[(ActorRef, GetShardHome)] = Set.empty
```
Comment (fsat) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: I still haven't figured out how to specifically test for non-duplicated `GetShardHome` message.

I did run the `validatePullRequest` before creating this PR which invokes `akka-cluster-sharding` and `akka-cluster-sharding-typed`, and all tests seems to pass.

If anyone has any hint or ideas on how to test for non-duplicated `GetShardHome` message, would you kindly share them with me please? Or maybe it's not necessary to specifically test for this due to existing coverage in the `ClusterShardingSpec.scala`?

Any advice will be greatly appreciated - thanks!

Comment (raboof) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Perhaps:
* create Coordinator
* send `GetShardHome`, validate response
* make it go into a state where it needs to update
* send 2 `GetShardHome`s
* send the update
* expect exactly 1 response (then `expectNoMessage`)

Would that be possible? Or is there anything that makes that scenario hard to reconstruct in a test case?

Comment (fsat) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: @raboof - sorry for the late response, I've been away for a couple of days.

I will try with your suggestion and report back here.

**Comment by patriknw** [comment]
Refs #25444

**Review by patriknw — COMMENTED** [review]
looks like a good approach

**Comment by fsat** [comment]
@patriknw - thanks for taking a look, Patrik! I really appreciate it. I am trying out suggestion from @raboof - once I got it done, I will send it for a review.

**Review by patriknw — APPROVED** [review]
LGTM, sorry for delays here. I can update this for latest master. Maybe it stalled on the testing, and I think it's covered enough (implicitly) by all other tests. Conceptually it's the same as sending a retry so should be a safe change.

**Comment by patriknw** [comment]
Continued in https://github.com/akka/akka/pull/26923

**Pull request closed without merging**